<launch>

  <!-- All basic features like loading the fundamental parameters, the robot description parameters etc. are imported from basic.launch, 
  so refer to that -->
  <arg name="gazebo_gui" default="false" />
  <arg name="set_joints" default="true" />
  

  <include file="$(find moto_project)/launch/basic.launch">
    <arg name="set_joints" value="$(arg set_joints)" />
    
    <arg name="model_name" value="sia5_with_rgbd_camera.urdf.xacro"/>
    <arg name="gazebo_gui" default="$(arg gazebo_gui)" />
  </include>

  <!-- parameters to show or not optional things -->
  <arg name="rviz_gui" default="True" />
  <arg name="set_debug" default="False" />
  <arg name="view_segmentation" default="False" />
  <param name="classes" textfile="/home/alessio/progetto_robotica/DenseFusion/datasets/my_dataset/dataset_config/classes.txt" />
  <param name="yolact_model" value="/home/alessio/progetto_robotica/yolact_net/weights/yolact_resnet50_399_400000.pth" />

  <!-- topic to publish the images of the camera -->
  <arg name="image_topic" default="/sia5/rgbd/color/image_raw" />
  <arg name="depth_topic" default="/sia5/rgbd/depth/image_raw" />
  <arg name="point_cloud_topic" default="/sia5/rgbd/depth/points" />
  <arg name="segmentation_topic" default="/sia5/rgbd/segmentation_result" />
  <arg name="mask_topic" default="/sia5/rgbd/mask_result" />
  <arg name="mask_image_topic" default="/sia5/mask_image" />

  <!-- parameters to commond use -->
  <param name="image_topic_name" value="$(arg image_topic)" />
  <param name="depth_topic_name" value="$(arg depth_topic)" />
  <param name="point_cloud_topic_name" value="$(arg point_cloud_topic)" />
  <param name="segmentation_topic_name" value="$(arg segmentation_topic)" />
  <param name="mask_topic_name" value="$(arg mask_topic)" />
  <param name="mask_image_topic_name" value="$(arg mask_image_topic)" />
  <param name="debug" value="$(arg set_debug)" />

  <!-- models parameters -->
  <arg name="x_table_offset" default="0.2" />
  <arg name="x_can_offset" default="0.5" />

  <arg name="z_table_offset" default="0.25" />
  <arg name="z_can_offset" default="0.05" />
  
  <arg name="x_can_position" default="$(eval arg('x_table_offset') + arg('x_can_offset'))" />
  <arg name="z_can_position" default="$(eval arg('z_table_offset') + arg('z_can_offset'))" />

   <!-- spawn models -->
  <node name="urdf_spawner2" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen" 
  args="-file $(find motoman_description)/urdf/objects/coffee/coffee_can.urdf -urdf -model coffee_can  -x $(arg x_can_position) -z $(arg z_can_position) -wait table" /> 

  <node name="urdf_spawner3" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen" 
  args="-file $(find motoman_description)/urdf/objects/large_plane/large_plane.sdf -sdf -model table -x $(arg x_table_offset) -z $(arg z_table_offset)" /> 

   <!-- spawn segmentation node -->
  <node name="seg_node" pkg="moto_project" type="yolact_node.py" respawn="true" />
  <node name="mask_node" pkg="moto_project" type="mask_to_img.py" respawn="true" if="$(arg set_debug)"/>

  <!-- spawn nodes to visualize depth, rgb and optional the segmentation result -->
  <node name="image_viewer" pkg="rqt_image_view" type="rqt_image_view" args="$(arg image_topic)" />
  <node name="depth_viewer" pkg="rqt_image_view" type="rqt_image_view" args="$(arg depth_topic)" />
  <node name="segmentation_viewer" pkg="rqt_image_view" type="rqt_image_view" args="$(arg segmentation_topic)" if="$(arg view_segmentation)"/>
  <node name="mask_viewer" pkg="rqt_image_view" type="rqt_image_view" args="$(arg mask_image_topic)" if="$(arg set_debug)"/>

  <!--
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find moto_project)/rviz/rgbd_config.rviz" if="$(arg rviz_gui)"/> -->
   <node name="rviz" pkg="rviz" type="rviz" args="-d $(find moto_project)/rviz/estimation.rviz" if="$(arg rviz_gui)"/> 
  
  <!-- ros_control motoman launch file -->
</launch>
